dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -185,13 +185,7 @@ def sign_up_google_oauth2
       @user = user
       silent_takeover @user, auth_hash
       sign_in_user
-    elsif (looked_up_user = User.find_by_email_or_hashed_email(user.email))
-      email_already_taken_redirect \
-        provider: AuthenticationOption::GOOGLE,
-        found_provider: looked_up_user.provider,
-        email: user.email
```

Comment:
This code is unreachable because:

It falls beneath the `if` clause above:

https://github.com/code-dot-org/code-dot-org/blob/90b09b5598f152b87da9719d8fb90790b8cedb2a/dashboard/app/controllers/omniauth_callbacks_controller.rb#L184

Which is defined thus:

https://github.com/code-dot-org/code-dot-org/blob/90b09b5598f152b87da9719d8fb90790b8cedb2a/dashboard/app/controllers/omniauth_callbacks_controller.rb#L438-L443

Let's break these cases into named clauses to help work out the logic:

- `A: AuthenticationOption::SILENT_TAKEOVER_CREDENTIAL_TYPES.include?(auth_hash.provider.to_s)`
- `B: User.find_by_email_or_hashed_email(oauth_user.email)`
- `C: !oauth_user.persisted?`

So we can express our branching logic this way:
```
if A && B && C
  silent takeover and sign in
elsif B
  case we want to eliminate
else
  register new user
```

Now let's simplify some things:

- Because this is in the Google-specific callback, `A` is always `true` and can simply be eliminated from this expression.
- `B` and `C` do not vary independently: `User.from_omniauth` will persist a google user if and only if that user's email is not already taken. (This is not true for all providers - more on that later). Therefore, we can say if `B` is true, then `C` must also be true. (If B is false then C may be true or false).  Therefore `B && C` depends entirely on the value of `B` and `C` can be eliminated from this expression.

That reduces our logic to:
```
if B
  silent takeover and sign in
elsif B
  case we want to eliminate
else
  register new user
```

We can safely eliminate the second case.